discovery/manager.go

```diff
@@ -0,0 +1,288 @@
+func NewManager(ctx context.Context, logger log.Logger) *Manager {
+	return &Manager{
+		ctx:            ctx,
+		logger:         logger,
+		actionCh:       make(chan func()),
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		targets:        make(map[string]map[string][]*config.TargetGroup),
+		discoverCancel: []context.CancelFunc{},
+	}
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+type Manager struct {
+	ctx            context.Context
+	logger         log.Logger
+	syncCh         chan map[string][]*config.TargetGroup // map[targetSetName]
+	actionCh       chan func()
+	discoverCancel []context.CancelFunc
+	targets        map[string]map[string][]*config.TargetGroup // map[targetSetName]map[providerName]
+}
+
+// Run starts the background processing
+func (m *Manager) Run() error {
+	for {
+		select {
+		case f := <-m.actionCh:
+			f()
+		case <-m.ctx.Done():
+			m.cancelDiscoverers()
+			return m.ctx.Err()
+		}
+	}
+}
+
+// SyncCh returns a read only channel used by all Discoverers to send target updates.
+func (m *Manager) SyncCh() <-chan map[string][]*config.TargetGroup {
+	return m.syncCh
+}
+
+// ApplyConfig removes all running discovery providers and starts new ones using the provided config.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	err := make(chan error)
+	m.actionCh <- func() {
+		m.cancelDiscoverers()
+		for _, scfg := range cfg.ScrapeConfigs {
+			for provName, prov := range m.providersFromConfig(scfg.ServiceDiscoveryConfig) {
+				m.startProvider(scfg.JobName, provName, prov)
+			}
+		}
+		close(err)
+	}
+
+	return <-err
+}
+
+func (m *Manager) startProvider(jobName, provName string, worker Discoverer) {
+	ctx, cancel := context.WithCancel(m.ctx)
+	updates := make(chan []*config.TargetGroup)
+
+	m.discoverCancel = append(m.discoverCancel, cancel)
+
+	go worker.Run(ctx, updates)
+	go func(provName string) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case tgs, ok := <-updates:
+				// Handle the case that a target provider exits and closes the channel
+				// before the context is done.
+				if !ok {
+					return
+				}
+				m.syncCh <- m.mergeGroups(jobName, provName, tgs)
+			}
+		}
+	}(provName)
+}
+
+func (m *Manager) cancelDiscoverers() {
+	for _, c := range m.discoverCancel {
+		c()
+	}
+	m.targets = make(map[string]map[string][]*config.TargetGroup)
+	m.discoverCancel = []context.CancelFunc{}
+}
+
+// mergeGroups adds a new target group for a given discovery provider and returns all target groups for a given target set
+func (m *Manager) mergeGroups(tsName, provName string, tg []*config.TargetGroup) map[string][]*config.TargetGroup {
+	tset := make(chan map[string][]*config.TargetGroup)
+
+	m.actionCh <- func() {
+		if m.targets[tsName] == nil {
+			m.targets[tsName] = make(map[string][]*config.TargetGroup)
+		}
+		m.targets[tsName][provName] = []*config.TargetGroup{}
```

Comment:
Same as above. no need to instantiate an empty slice unless for pre-allocation. This line can just be dropped.